clang/runtime/dpct-rt/include/dpct/device.hpp

```diff
@@ -732,6 +741,31 @@ class dev_mgr {
     select_device(selected_device_id);
   }
 
+  /// Update the device stack for the current thread id
+  void push_device(unsigned int id) {
+    auto tid = get_tid();
+    auto it = _thread2dev_map.find(tid);
+    if (it == _thread2dev_map.end())
+      _thread2dev_map.insert({tid, std::stack<unsigned int>({id})});
+    else
+      it->second.push(id);
+  }
+
+  /// Remove the device from top of the stack if it exist
+  unsigned int pop_device() {
+    auto it = _thread2dev_map.find(get_tid());
+    if (it == _thread2dev_map.end())
+      return -1;
```

Comment:
Should throw exception instead of returning `-1`.

Reply:
`cuCtxPopContext` returns as non-zero error code but let the program continue. An sample code is [here](https://github.com/oneapi-src/SYCLomatic/pull/2109#discussion_r1662329017).

Shouldn't we maintain the same behavior?

Reply:
The returned error code can be checked in the CUDA code, but there is no such mechanism in SYCL. So, I suggest throwing the exception.

Reply:
Done